CHIP-0025: Chialisp Message Conditions

danieljperry opened the pull request.

danieljperry commented:
This CHIP is now a `Draft`. It introduces a new set of Chialisp conditions to ensure that an exact set of coins is spent together. This CHIP is still in an early stage, where the conditions are described, but not yet implemented in code. Feel free to leave a review here, and follow along as more updates are added to this PR in the coming weeks.

danieljperry commented:
We have made some rather large updates to the message conditions and how they work. There are now only two conditions: one for sending messages, and one for receiving them. They use a six-bit bitmask to communicate the type of message to send/receive. Let us know what you think!

danieljperry commented:
On March 12, we will have a public Zoom call to discuss this CHIP. See the #chips channel in [our discord](https://discord.com/invite/chia) for more details.

danieljperry commented:
This CHIP is being implemented in [PR #430](https://github.com/Chia-Network/chia_rs/pull/430) of the [chia_rs repository](https://github.com/Chia-Network/chia_rs).

arvidn commented on CHIPs/chip-0025.md:
```diff
@@ -0,0 +1,156 @@
+| Puzzle        | 010  | `<puzzle hash>`                    |
+| Parent-Puzzle | 110  | `<parent coin ID>` `<puzzle hash>` |
+| Parent-Amount | 101  | `<parent coin ID>` `<amount>`      |
+| Puzzle-Amount | 011  | `<puzzle hash>` `<amount>`         |
```
since the bitmasks `000` and `001` are still supported in consensus mode. I think those should be included here as well. They take no argument and `<amount>` respectively

danieljperry commented:
Please see this Youtube video of our discussion of this CHIP with the community:
https://youtu.be/zCuP_mo3Z7s

danieljperry commented:
This CHIP is now in `Review`. Please make your final reviews soon. Note that the `chia-blockchain` PR and fork activation height are still pending.

danieljperry commented:
A couple updates for this CHIP:
1. We have decided to allow all eight bitmask combinations in both mempool mode and in the consensus. Previously we had excluded `amount` and `none` from the mempool, but allowed them in the consensus. We are adding these combinations because there are likely legitimate use cases for them, and because it should be less confusing when everything is allowed.
2.  We have set `5 550 000` as the tentative block height for the soft fork.

danieljperry commented:
We have decided to push back the fork point for this CHIP to block `5 650 000`.

danieljperry commented:
One minor change -- we changed the name of the fork to `soft fork 4`. The previous name, `soft fork 3`, had been planned in a different CHIP that was withdrawn. We renamed this fork in order to avoid confusion between the two.

Please get your final reviews in, as this CHIP will be moved to `Last Call` soon.

danieljperry commented:
We have adjusted the activation height for this CHIP to block `5 716 000`.

danieljperry commented:
This CHIP is now in `Last Call`. If no changes are required for the next two weeks, it will be moved to `Final`.

danieljperry commented:
This CHIP is now `Final`. No further changes are allowed. The conditions introduced in this CHIP will become available after the soft fork activates, which is expected to occur around July 31, 2024. The exact block height of the activation is `5 716 000`.

hoffmang9 reviewed (approved):
Lgtm
